petastorm/reader_impl/shuffling_buffer.py

```diff
@@ -0,0 +1,168 @@
+    def can_retrieve(self):
+        """Checks the state of the buffer and returns whether an item can be removed from the buffer..
+
+        :return: A boolean indicating whether an item can be returned from the buffer at the time.
+        """
+        pass
+
+    @abc.abstractproperty
+    def size(self):
+        """Returns the number of elements currently present in the buffer.
+
+        :return: number of elements currently present in the buffer
+        """
+        pass
+
+    @abc.abstractmethod
+    def finish(self):
+        """Call this method when no more `enqueu_many` calls will be made.
+
+        :return: number of elements currently present in the buffer
+        """
+        pass
+
+class NoopShufflingBuffer(ShufflingBufferBase):
+    """A 'no-operation' (noop) implementation of a shuffling buffer. Useful in cases where no shuffling is desired, such
+    as test scenarios or iterating over a dataset in a predeterministic order.
+    """
+    def __init__(self):
+        self.store = []
+
+    def add_many(self, items):
+        self.store.extend(items)
+
+    def retrieve(self):
+        return self.store.pop(0)
+
+    def can_retrieve(self):
+        return len(self.store) > 0
+
+    def can_add(self):
+        return True
+
+    @property
+    def size(self):
+        return len(self.store)
+
+    def finish(self):
+        pass
+
+class RandomShufflingBuffer(ShufflingBufferBase):
+    """
+    A random shuffling buffer implementation. Items can be added to the buffer and retrieved in a random order.
+    """
+    def __init__(self, shuffling_buffer_capacity, min_after_retrieve, extra_capacity=1000):
+        """Initializes a new ShufflingBuffer instance.
+
+        Items may be retrieved from the buffer once `min_after_dequeue` items were added to the queue
+        (indicated by `can_retrieve`).
+
+        Items may be added to the buffer as long as the amount of items in the buffer does not exceed
+        `shuffling_queue_capacity`.
+
+        The amount of items in the buffer may grow above `shuffling_buffer_capacity` (due to a call to `add_many`
+        with a list of items), but must remain under `extra_queue_capacity`.
+
+        :param shuffling_buffer_capacity: Items may be added to the buffer as long as the amount of items in the
+        buffer does not exceed the value of `shuffling_queue_capacity`.
+        :param min_after_retrieve: The amount of items in the buffer may grow above `shuffling_buffer_capacity`
+        (due to a call to `add_many` with a list of items), but must remain under `extra_queue_capacity`
+        """
+        self._extra_capacity = extra_capacity
+        # Preallocate the shuffling buffer.
+        self._items = [None] * (shuffling_buffer_capacity + self._extra_capacity)
+        self._shuffling_queue_capacity = shuffling_buffer_capacity
+        self._min_after_dequeue = min_after_retrieve
+        self._size = 0
+        self._done_adding = False
+
+    def add_many(self, items):
```

Comment:
This doesnt look thread safe. Do we need it to be thread safe? (I think the answer is no since the main thread will "poll" and move things between queues but I wanted to be sure. Also it would be useful to document in the baseclass that a shuffling queue does not need to be thread safe)

Reply:
Added a comment that this is not a threadsafe class to the base class.